src/main/java/games/strategy/engine/framework/startup/login/V2Authenticator.java

```diff
@@ -0,0 +1,222 @@
+    }
+
+    try {
+      return Base64.getDecoder().decode(value);
+    } catch (final IllegalArgumentException e) {
+      throw new AuthenticationException(String.format("malformed value for property '%s'", name), e);
+    }
+  }
+
+  private static byte[] digest(
+      final String password,
+      final byte[] salt,
+      final byte[] nonce)
+      throws GeneralSecurityException {
+    final Mac mac = Mac.getInstance("HmacSHA512");
+    mac.init(newSecretKey(password, salt));
+    return mac.doFinal(nonce);
+  }
+
+  private static SecretKey newSecretKey(final String password, final byte[] salt) throws GeneralSecurityException {
+    // 20,000 iterations was empirically determined to provide good performance on circa 2014 hardware for
+    // PBKDF2WithHmacSHA512.
+    //
+    // https://security.stackexchange.com/questions/3959/recommended-of-iterations-when-using-pkbdf2-sha256
+    final int iterationCount = 20_000;
+
+    // Per RFC 4868, the key length for HmacSHA512 should be between 512 bits (L) and 1,024 bits (B). However, some
+    // references insist that, in the context of current computing power to break crypto, 128 bits should be
+    // sufficient.
+    //
+    // https://security.stackexchange.com/a/96176/136686
+    final int keyLengthInBits = 512;
+
+    final SecretKeyFactory secretKeyFactory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA512");
+    final KeySpec keySpec = new PBEKeySpec(password.toCharArray(), salt, iterationCount, keyLengthInBits);
+    return secretKeyFactory.generateSecret(keySpec);
+  }
+
+  /**
+   * Authenticates a client based on a previously-sent challenge.
+   *
+   * <p>
+   * If this method does not throw an exception, the client is considered to have authenticated successfully. Otherwise,
+   * the exception will contain the details of the failure. No information from the exception (including the message)
+   * should be returned to the client.
+   * </p>
+   *
+   * @param password The game password; must not be {@code null}.
+   * @param challenge The challenge sent to the client node; must not be {@code null}.
+   * @param response The response received from the client node; must not be {@code null}.
+   *
+   * @throws AuthenticationException If the client did not authenticate successfully or if an error occurs while parsing
+   *         the challenge or response.
+   */
+  static void authenticate(
+      final String password,
+      final Map<String, String> challenge,
+      final Map<String, String> response)
+      throws AuthenticationException {
+    final byte[] nonce = decodeRequiredProperty(challenge, ChallengePropertyNames.NONCE);
+    final byte[] salt = decodeRequiredProperty(challenge, ChallengePropertyNames.SALT);
+    final byte[] actualDigest = decodeRequiredProperty(response, ResponsePropertyNames.DIGEST);
+
+    try {
+      final byte[] expectedDigest = digest(password, salt, nonce);
+      if (!MessageDigest.isEqual(expectedDigest, actualDigest)) {
```

Comment:
+1 What I did was encrypting the recieved message with the generated key which leads to exceptions if the password is wrong.
This is a better approach and resistant against timing attacks, although those shouldn't be that big of a problem because the challenge changes every attempt.